drafts/swift/redux-in-swift-pt-1.md

```diff
@@ -0,0 +1,280 @@
+class Store<State> {
+    private(set) var state: State! // This `!` will be explained later
+}
+```
+
+The `state` is marked as a `var` because:
+
+the *State* in a *Store* can be changed by dispatching *Actions* to a *Store*.
+
+
+The `state` is marked as `private(set)` because:
+
+- the *State* in a *Store* should be read-only and
+- "The only way to change the state is to emit an action, an object describing what happened."
+
+[Redux Doc: Three Principles](http://redux.js.org/docs/introduction/ThreePrinciples.html)
+
+### Reducer and Action
+
+A *Store* need a *Reducer* to generate new *State*s from the current *State* and an Action, which means:
+
+- A *Store* need to hold onto a *Reducer*
+- A *Reducer*
+    - takes a *State* and a *Action* and
+    - returns a new *State*
+- An *Action* describes what happen**ed** which needs to differentiate from each other while it may carry some data
+    - An *Action* can be just an empty Struct or an Enum as demonstrated in later sections
+
+Essentially, a *Reducer* in Swift can be defined as a function type.
+
+
+```swift
+protocol ActionType { }
+
+class Store<State> {
+    // ... previous code ...
+    typealias Reducer = (state: State?, action: ActionType) -> State
+    final let reducer: Reducer
+    
+    init(with reducer: Reducer) {
+        self.reducer = reducer
+    }
+}
+```
+
+#### Things to Note about Reducers
+
+- They should be pure: No side-effects. No API calls. etc.
+- They take `nil` as `State` which implies the default initial state is defined in a *Reducer* rather than a *Store*.
+    - This is partially why we need that `!` when defining the `state` in the `Store`.
+- They should return the current *State* if they are given an *Action* they cannot handle.
+
+### Dispatching Actions and Notifying Subscribers
+
+An *Action* can be dispatched to a *Store* to change its *State*. After a *Store*'s *State* is changed, the *Store* should notify all of its *Subscribers*. Thus,
+
+- A *Subscriber* is a function takes a *Store*
+- *Subscribers* can subscribe to a *Store* (will be explained in the next section)
+- A *Store* need to hold onto an array of *Subscribers*
+- After a *Store*'s *State* changes, the *Store* need to notify the change to all of its *Subscribers*
+
+```swift
+class Store<State> {
+    // ... previous code ...
+    typealias Subscriber = (store: Store) -> ()
+    final var subscribers = [Subscriber]()
+
+    final func dispatch(action: ActionType) {
+        self.state = reducer(state: state, action: action)
+        subscribers.forEach {
+            $0(store: self)
+        }
+    }
+}
+```
+
+Now the generic class `Store` is almost usable except a few pieces missing. Example usage:
+
+```swift
+struct Increase: ActionType { }
+
+let counterStore = Store<Int>.init { (state: Int?, action: ActionType) -> Int in /* ... */ }
+
+counterStore.dispatch(Increase())
+```
+
+If you want to run above snippet with minimum change, replace the `/* ... */` with `return 0`.
```

Comment:
*If you want to run the snippet above at this point, you'll need to replace the...